File: engine/renderer/camera.cpp
//
// Created by Nemo li on 2023/12/7.
//

#include "camera.h"
#include <memory>
#include <glm/gtc/matrix_transform.hpp>
#include <glad/glad.h>
#include <rttr/registration>
#include "component/game_object.h"
#include "component/transform.h"

RTTR_REGISTRATION//注册反射
{
    rttr::registration::class_<Camera>("Camera")
            .constructor<>()(rttr::policy::ctor::as_raw_ptr);
}

std::vector<Camera *> Camera::all_camera_;
Camera *Camera::current_camera_;

Camera::Camera()
        : Component(), clear_color_(49.f / 255, 77.f / 255, 121.f / 255, 1.f),
          clear_flag_(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT), depth_(0),
          culling_mask_(0x01) {
    //默认获取现有Camera最大depth，设置当前Camera.depth +1
    if (all_camera_.size() > 0) {
        unsigned char max_depth = all_camera_.back()->depth();
        depth_ = max_depth + 1;
    }
    all_camera_.push_back(this);
}


Camera::~Camera() {
    auto iter = std::find(all_camera_.begin(), all_camera_.end(), this);
    if (iter != all_camera_.end()) {
        all_camera_.erase(iter);
    }
}

void Camera::SetView(const glm::vec3 &cameraForward, const glm::vec3 &cameraUp) {
    auto transform = game_object()->GetComponent<Transform>();
    view_mat4_ = glm::lookAt(transform->position(), cameraForward, cameraUp);
}

void Camera::SetPerspective(float fovDegrees, float aspectRatio, float nearClip, float farClip) {
    projection_mat4_ = glm::perspective(glm::radians(fovDegrees), aspectRatio, nearClip, farClip);
}

void Camera::SetOrthographic(float left, float right, float bottom, float top, float z_near, float z_far) {
    projection_mat4_ = glm::ortho(left, right, bottom, top, z_near, z_far);
}

void Camera::Clear() {
    glClear(clear_flag_);
    glClearColor(clear_color_.r, clear_color_.g, clear_color_.b, clear_color_.a);
}

void Camera::set_depth(unsigned char depth) {
    if (depth_ == depth) {
        return;
    }
    depth_ = depth;
    Sort();
}

void Camera::Sort() {
    std::sort(all_camera_.begin(), all_camera_.end(), [](Camera *a, Camera *b) {
        return a->depth() < b->depth();
    });
}

void Camera::Foreach(std::function<void()> func) {
    for (auto iter = all_camera_.begin(); iter != all_camera_.end(); iter++) {
        current_camera_ = *iter;
        current_camera_->Clear();
        func();
    }
}

File: engine_lua/renderer/camera.lua
﻿require("lua_extension")

--- 相机
--- @class Camera : CppComponent
Camera = class("Camera", Component)

function Camera:ctor()
    Camera.super.ctor(self)
end

function Camera:InitCppComponent()
    ---@type Cpp.Camera
    self.cpp_component_instance_ = Cpp.Camera() --创建对应的C++实例
end

--- 设置相机位置 朝向
--- @param camera_forward vec3 @相机朝前方向
--- @param camera_up vec3 @相机朝上方向
function Camera:SetView(camera_forward, camera_up)
    self.cpp_component_instance_:SetView(camera_forward, camera_up)
end

---@param fov_degrees number @相机视野 可视角度
---@param aspect_ratio number @宽高比
---@param near_clip number @近裁剪面
---@param far_clip number @远裁剪面
function Camera:SetPerspective(fov_degrees, aspect_ratio, near_clip, far_clip)
    self.cpp_component_instance_:SetPerspective(fov_degrees, aspect_ratio, near_clip, far_clip)
end

function Camera:view_mat4()
    return self.cpp_component_instance_:view_mat4()
end

function Camera:projection_mat4()
    return self.cpp_component_instance_:projection_mat4()
end

--- 设置正交相机矩阵
--- @param left number
--- @param right number
--- @param bottom number
--- @param top number
--- @param z_near number
--- @param z_far number
function Camera:SetOrthographic(left, right, bottom, top, z_near, z_far)
    self.cpp_component_instance_:SetOrthographic(left, right, bottom, top, z_near, z_far)
end

--- 设置清屏颜色
--- @param r number
--- @param g number
--- @param b number
--- @param a number
function Camera:set_clear_color(r, g, b, a)
    self.cpp_component_instance_:set_clear_color(r, g, b, a)
end

--- 设置刷帧清屏内容种类
--- @param clear_flag number
function Camera:set_clear_flag(clear_flag)
    self.cpp_component_instance_:set_clear_flag(clear_flag)
end

--- 刷帧清屏
function Camera:Clear()
    self.cpp_component_instance_:Clear()
end

function Camera:depth()
    return self.cpp_component_instance_:depth()
end

--- 设置 depth，触发相机排序
--- @param depth number @相机深度，多相机按从后往前排序，按depth从小往大遍历，depth越大的越上层。
function Camera:set_depth(depth)
    self.cpp_component_instance_:set_depth(depth)
end

function Camera:culling_mask()
    return self.cpp_component_instance_:culling_mask()
end

function Camera:set_culling_mask(culling_mask)
    self.cpp_component_instance_:set_culling_mask(culling_mask)
end

function Camera:camera_use_for()
    self.cpp_component_instance_:camera_use_for()
end